Kcena 4주차 과제1 [todo app에서 redux 적용하기] 제출합니다.

Kwakcena opened the pull request:
첫 시작을 어떻게 해야 할지 감이 안잡혔습니다. 제출이 너무 늦었네요 ㅠ. 😭

강의는 `redux`를 설명해야 되서 `redux`의 `reducer` 부분을 구현한 다음 test를 만든 것이죠? 
이와 반대로 `reducer.test.js`에서 테스트를 작성하고 구현하는 방식으로 과제를 수행해봤습니다.

하나의 기능에 대해 `RED - GREEN - REFACTOR` 순서를 지키려고 노력해봤는데, 잘 안되네요.
`addTask`를 하고 다 했다 생각해서 `deleteTask`를 하다가 다시 `addTask`에서 안한게 생각나서 넘어가기도 하고... 이러면 안되는거죠?
한가지 기능에 대한 테스트를 집중적으로 완료한 뒤에 넘어가는게 맞다고 생각하는데, 테스트 작성하면서 빠트린것은 없는지
신중해야 된다는 느낌이 들었습니다.

아직 미완성이지만 너무 졸려서 진행한 부분까지 제출하겠습니다.

Kwakcena committed:
App component의 통합 테스트 작성

App 컴포넌트는 상태를 변경할 수 있는 기능을 포함하고 있다.
- 할 일 입력하기
- 할 일 추가하기
- 할 일 삭제하기

이 세 가지 구현에 대해 테스트를 작성했다.

어떻게 시작해야 할 지 몰라서 App 테스트를 작성한 뒤 redux를 적용하면서
테스트를 고치는 방향으로 잡았다.

Kwakcena committed:
App 컴포넌트 안에서 관심사 분리하기

App 컴포넌트 내부에 상태를 변경하는 기능들이 담겨있고, 화면을 보여주는
부분까지 포함하고 있다.

상태를 관리하는 기능들을 밖으로 빼서 서로의 관심사를 분리한다.

Kwakcena committed:
Redux 설치

Kwakcena committed:
Add store.js

리덕스를 적용하기 위해 store.js 파일을 만들었다.

리덕스에서 store는 상태를 변경해주는 reducer와 상태값을 갖는 객체다.

Kwakcena committed:
index.jsx에 Provider 적용

Provider를 적용해서 하위 컴포넌트들이 store를 사용할 수 있도록 한다.

Kwakcena committed:
Action creaters 만들기

상태를 변화시키기 위해서는 dispatch를 이용해 action을 전달해야 한다. 이
action을 발생시키는 action creater 들을 모아놓은 actions.js를 만든다.

Kwakcena committed:
Reducer.js 파일 생성

이제 action에 따라 상태를 변경시켜 줄 reducer 함수를 만들고 테스트를
추가하며 하나씩 구현해보자.

Kwakcena committed:
[RED] reducer updateTaskTitle 테스트 코드 작성

updateTaskTitle에 대한 테스트 코드를 작성.

reducer에 updateTaskTitle에 대한 구현이 되어 있지 않아서 원하는
taskTitle을 기대할 수 없음.

Kwakcena committed:
[GREEN] updateTaskTitle 기능 구현

이전 상태와 action을 넣어주면 새로운 상태를 반환한다. 새로운 상태에는
입력한 taskTitle 값이 포함되어 반환된다.

Kwakcena committed:
[Refactor] actions의 사용, Inline Variable

- actions 에 정의된 updateTaskTitle action creater를 사용.
- 이전 상태값인 previousState 변수에 대해 Inline variable refactoring

see also : https://refactoring.com/catalog/inlineVariable.html

Kwakcena committed:
[RED] addTask test

할 일을 추가하면 새로운 상태의 newId는 +1이 되어야 한다.

Kwakcena committed:
[GREEN] addTask test

addTask에 대한 reducer를 추가한다. 이제 id 뿐만 아니라 실제 tasks와
taskTitle이 추가되는지 확인해야 한다.

Kwakcena committed:
[Refactor] tasks 확인, 추가 이후 초기화 테스트 추가

할 일이 추가된 후 tasks에 추가되어 길이가 0이 아닌지 확인하고, 추가된
뒤로 taskTitle이 빈 값으로 변하는지 테스트를 추가한다.

Kwakcena committed:
[Refactor] Inline Variable refactoring

Kwakcena committed:
[RED] deleteTask

할 일 목록 삭제 구현 test

Kwakcena committed:
[GREEN] deleteTask

할 일이 삭제되면 tasks의 항목 길이는 0이 된다.

Kwakcena committed:
[Refactor] addTask

addTask에서 중복된 부분을 refactoring 한다.

Kwakcena committed:
[RED] addTask

addTask에서 만약 입력창에 아무것도 쓰지 않고 추가버튼을 눌렀다면
추가되지 않아야 하는데, 현재는 추가되도록 구현되어 있다. 이를 수정한다.

Kwakcena committed:
[GREEN] addTask

할 일이 없으면 아무것도 추가되지 않도록 한다.

Kwakcena committed:
[Refactor] addTask 테스트를 좀더 명확하게

addTask에서 할 일이 추가되면 tasks 배열의 길이가 0이 아닌지로 테스트를
판단했는데, 실제로 tasks 내부에 추가되었는지 확인하는 테스트 코드를
작성해서 명확하게 했다.

Kwakcena committed:
[Refactor] reducer test

context는 with, without의 개념에 사용되는 것 같아서 describe로 바꿨다.

Kwakcena committed:
[Refactor] reducer에서 if 문 없애기

계산기 과제에서 if문 없이 객체에 operator 를 이용해서 각각의 함수를
호출해 사용하던 방법을 적용했다.

Kwakcena committed:
잘못된 코드 수정

createStore 를 react-redux에서 가져오던 걸 redux로 수정

Kwakcena committed:
[막힌 부분] 초기상태, if문 없애기

- if문을 없애기 위해 작성한 코드가 실제 npm start를 돌렸을 때 에러를
  발생시켰습니다.

처음 앱이 실행될 때 createStore 하는 과정에서 reducer 함수를 실행시키면
action.type 에 이상한 값이 있어서 제대로 된 객체의 key를 찾을 수 없어
발생한 에러였습니다. 이 부분을 해결하기 위해 handler에 제대로 된 값이
들어갔을 경우 해당 handler의 함수를 실행시키도록 했지만...이런 부분은
TDD로 어떻게 잡아야 하나요?

- reducer의 초기 상태에 대한 테스트

reducer의 state 기본 값이 initState가 들어가는 부분은 테스트로 어떻게
해야 할지 모르겠습니다.

wholemann reviewed (commented):
'하나의 테스트에 대한 구현을 완벽하게 완료하기 전에

다른 구현으로 넘어가선 안 된다' 라고 너무 강박적으로 생각하진 마세요:)

어차피 구현하다 다른 부분에 문제가 생기면 그건 자동화된 테스트에서 알려줄 겁니다.

그러면 그 때 문제를 해결하면 되는 거죠.

TDD는 피드백 주기를 굉장히 짧게 가져가서 문제 발견을 굉장히 자주 할 수 있게 됩니다.

구현을 하다 보면 당연히 Test에서 놓친 부분이 생각나는게 정상입니다.

그러면 Test에 대한 수정이 필요하구요. 수정된 Test를 통과시키기 위해선

의존성이 있는 다른 구현체들을 수정해야 될 수 도 있습니다.

중요한 건 이러한 사실을 빨리 발견할 수 있다는 거죠. 이게 TDD의 힘입니다.

찌금 잘하고 계시니 잘 마무리 해주시면 될 것 같아요ㅎㅎ

Kwakcena committed:
App 컴포넌트 테스트를 위한 __mocks__ 추가

App 컴포넌트에서 redux를 사용하기 전, redux를 사용하는 테스트 코드
작성이 필요하다.

테스트 코드에서 useSelector, useDispatch를 사용하기 위해 __mocks__
폴더를 만들어 가짜 함수를 넣어준다.

Kwakcena committed:
[RED] App component test

App 컴포넌트는 리덕스를 사용하여 구현에 대한 관심사는 갖고 있지 않다.
따라서 그에 맞게 테스트의 내용도 수정되어야 한다.

useSelector로 tasks 목록을 가져와서 해당 목록의 title 값이 화면에
보이는지 테스트를 수행한다.

Kwakcena committed:
[GREEN] 실제 코드에서 Redux를 사용하도록 작성

실제 App 컴포넌트에서도 useSelector와 useDispatch를 사용하도록 코드를
작성했다.

신기한점은 App.test.jsx 에서 useDispatch를 사용하지 않았는데 dispatch가
함수가 아니라는 에러가 발생했었다.
그래서 __mocks__ 에 useDispatch에 대한 가짜 함수를 넣어줬더니
해결되었다.
가져와서 사용하지도 않았는데 왜 에러가 나타난걸까?

Kwakcena committed:
[GREEN] 할 일이 없을 경우에 대한 테스트 추가

생각난 테스트를 뒤늦게 추가

Kwakcena committed:
[Refactor] describe-context-it 구조 사용

context를 추가함

Kwakcena committed:
[RED] ListContainer 할 일 목록 보여주기

할 일 목록이 있을 때 목록을 보여주는 테스트 작성.

Kwakcena committed:
[GREEN] ListContainer 할 일 목록 보여주기

ListContainer는 taskTitle에 관심이 없고 tasks에 관심을 갖고 있으며, 할
일 목록을 지우는 기능과 관련있다.

Kwakcena committed:
[Refactor] ListContainer 할 일 없을 경우

할 일 없을 경우 테스트 추가

Kwakcena committed:
[Refactor] 다른 컴포넌트에서 ListConatiner 사용

ListContainer의 테스트가 완료 되어 사용할 수 있다는 판단 하에 다른
컴포넌트에 실제로 적용한다.

Kwakcena committed:
[RED] InputContainer test

InputContainer에서 할 일을 추가할 경우에 대한 테스트 작성

Kwakcena committed:
[GREEN] InputContainer 할 일 추가, value 테스트

value에 New Task가 잘 들어있는지 확인하고, 추가 버튼을 눌렀을 때
addTask가 호출되는지 확인한다.

Kwakcena committed:
[Refactor] page 컴포넌트 삭제, InputContainer 사용

Input 컨테이너를 사용하고 Page 컴포넌트를 삭제함.

Kwakcena committed:
[RED] reducer test - action이 없을 경우

action이 정의되어 있지 않은 경우 상태 변화는 없어야 한다.

wholemann commented on src/reducer.js:
```diff
@@ -0,0 +1,39 @@
+};
+
+export default function reducer(state = initialState, action) {
+  const handlers = {
```
이 부분은 reducers가 의미상 맞겠습니다.

그 외에는 훌륭하게 if문을 제거해주셨습니다.

Kwakcena replied:
`reducers` 로 고치면 하나 꺼내온 값에 대해서는 뭐라 써야 할까요? `reducer`는 이미 쓰고있고 어떤 말이 좋을지 모르겠네요 ㅠ
그런데 이 부분에서요
```javascript
const handler = handlers[action.type];
if(handler) {
    return handler(action.payload);
}
return state;
```

이 부분을 `return handlers[action.type](action.payload)` 라 하면 왜 에러가 나는지에 대해서는 이해를 못했어요.

wholemann replied:
defaultReducer를 하나 따로 만드시면 됩니다.

이 친구는 그냥 state만 반환하는 거죠. action도 들어오지 않게 만들면 되겠죠?

wholemann reviewed (commented):
역시 형조님 잘하셨어요.

특히 reducer에서 if문 제거는 상당히 좋았습니다.

계속 진행해서 언능 통과하시고 과제2 가시면 좋을 것 같아요ㅎㅎ

Kwakcena committed:
[GREEN] action의 기본값을 빈 객체로 설정

빈 객체로 설정하면 handler 객체에 없는 key에 접근하려 할 때 에러가
발생하지 않는다.

Kwakcena committed:
[Refactor] Rename variable

handlers 변수 이름을 reducers로

reducers의 결과물은 newState라는 이름을 사용.

Kwakcena committed:
[Refactor] InputContainer test coverage 100%

InputContainer에서 test 커버리지의 100% 달성.

기존의 테스트에서는 updateTaskTitle에 대한 테스트가 없었기 때문에
테스트로 인해 실행되는 handleChangeTitle 함수가 없었다.

따라서 이에 대한 테스트를 작성하여 커버리지 100퍼센트를 달성한다.

Kwakcena committed:
[Refactor] props 이름 수정, deleteTask test

ListContainer의 onClickDelete props가 이름이 잘못 되어 전달이 제대로
되지 않았다. 그래서 테스트가 계속 실패했다.

이를 고치고 ListContainer에 추가하여 테스트를 통과하였다.

Kwakcena committed:
[Refactor] state가 없는 경우 초기 상태 사용

state가 없는 경우 초기 상태를 사용한다는 테스트가 필요함.

그래서 initState가 노란색으로 칠해져 있었다.

wholemann reviewed (approved):
고생하셨습니다. Merge할게요.

과제2 후다닥 진행해주세요ㅎㅎㅎ
